### ISSUE TITLE
Exception: Amount precision is too high for currency.

### ISSUE THREAD

**Issue by lukeholder** [thephpleague/omnipay-common#143]
Using the Đồng (BND) currency, which has zero decimal places (zero minor units), returns an exception: 
"Amount precision is too high for currency." inside the `->getAmount()` method.

I believe this commit is causing the error:

https://github.com/thephpleague/omnipay-common/pull/70/commits/529b18fe6b3196b4606b28548cb97ee913eaa914

This line of code changed in the above commit is:
```
$decimal_count = strlen(substr(strrchr(sprintf('%.8g', $amount), '.'), 1));
```
with amount of VND 103500000, (an `int` of 103500000) returns `6` as the decimal_count, which is obviously incorrect.

After restoring the original line:
```
$decimal_count = strlen(substr(strrchr((string)$amount, '.'), 1));
```
with amount of VND 103500000, (an `int` of 103500000) returns `0` as the `$decimal_count`, which is the correct number of decimal places, and the transaction succeeds.

I don't quite understand the rationale for the above commit, hoping for experienced omnipay minds can assist.

/cc @judgej

**Comment by judgej** [comment]
Sorry - only just spotted this.

The original commits came from issues with different decimal accuracies being set for different PHP installations, meaning a simple cast to `(string)` gave different results on different installations. For example, some sites may give 1.66000 and others may give 1.659999999997 The `sprintf()` was meant to bypass that and make sure the accuracy (decimal digits) was the same for all installs.

**Comment by barryvdh** [comment]
What do you suggest for 3.x?

**Comment by judgej** [comment]
Not sure. I've always delegated money amounts to popular libraries with the assumption that they know how to handle this type of stuff better than me.

**Comment by barryvdh** [comment]
Well, in principle, we shouldn't touch the given value in most cases, unless we need to convert to integers at all.

In this case, we seem te be making a wrong decimal count, by casting it to max 8 decimals, see https://github.com/thephpleague/omnipay-common/pull/150

**Comment by barryvdh** [comment]
Would the fix using `number_format` and removing insignificant zeros work?

**Comment by judgej** [comment]
Yes, it should do. An `rtrim($string, '0')` should remove the treailing zeros, which surprisingly the PHP documentation does not explicitly state are added. It also doesn't mention the number is rounded.

**Comment by barryvdh** [comment]
Only thing I'm concerned about is the 'magic' 8 number. Is it good for everyone? Should it be number of decimals + 2? Should it depend on precision?

**Comment by barryvdh** [comment]
And should we throw an error at all? Downside of rounding is that it will depend on user input, so could fail depending on user input, right?

**Comment by kurund** [comment]
Hello,

Is there a work around for this issue? Thank you.
